src/core/ItemStack.ts

```diff
@@ -2,29 +2,43 @@ import type { NbtTag } from '../nbt/index.js'
 import { NbtCompound, NbtInt, NbtString } from '../nbt/index.js'
 import { Identifier } from './Identifier.js'
 
+export interface DefaultItemComponentProvider {
+	hasDefaultComponent(item: Identifier, component: Identifier): boolean
+	getDefaultComponent<T>(item: Identifier, component: Identifier): T | undefined
+}
+
+
 export class ItemStack {
 	constructor(
 		public id: Identifier,
 		public count: number,
-		public components: Map<string, NbtTag> = new Map(),
+		public components: Map<string, NbtTag | '!'> = new Map(),
+		public defaultProvider?: DefaultItemComponentProvider
```

Comment:
I think it might be a better idea to not have this, and require that library users always include the default components in the `ItemStack` that they want to render. I'm already doing this for misode.github.io (see https://github.com/misode/misode.github.io/blob/master/src/app/services/ResolvedItem.ts#L25)

Reply:
Flattening it does make sense, rather than doing that dynamically on load. I do think that should be in deepslate though, since its not a trivial thing to do. Also seems weird that you need to do things to the ItemStack gotten from `fromNbt` before being able to use it.

So maybe we could add the `defaultProvider` to `fromNbt` and do the flattening there?

Also with your current implementation you rely on the fact that `Identifier.parse` can parse the `!` as part of the namespace.

Reply:
We also somehow need a way to add the default components to all the items in the `minecraft:bundle_contents` component, since they are rendered by the `bundle/selected_item` item model and the `bundle/fullness` range dispatch property depends on the `max_stack_size` components.